src/stream/src/executor/source/executor_core.rs

```diff
@@ -38,7 +41,7 @@ pub struct StreamSourceCore<S: StateStore> {
     pub(crate) latest_split_info: HashMap<SplitId, SplitImpl>,
 
     /// Stores information of the splits.
-    pub(crate) split_state_store: SourceStateTableHandler<S>,
+    pub(crate) split_state_store: Arc<Mutex<SourceStateTableHandler<S>>>,
```

Comment:
Can we avoid introducing the lock? IMO we only need the `wait_epoch` call on the state store, without touching the internal fields of local states.

Reply:
The problem is we need to share the `SourceStateTableHandler<S>` between source executor and the worker. Although the worker only require share access, the source executor need mutable access. One improvement is to move the lock into the struct:
```
pub struct SourceStateTableHandler<S: StateStore> {
    pub state_store: Arc<Mutex<StateTable<S>>>,
}
```
WDUT?

Reply:
Can we directly call `wait_epoch` on the `StateStore`? It seems to be clonable.

Reply:
It seems not. It requires the  `<S as risingwave_storage::StateStore>::Local` to be cloneable.
```
error[E0599]: the method `clone` exists for struct `StateTableInner<S, BasicSerde, false, WatermarkBufferByEpoch<300>>`, but its trait bounds were not satisfied
   --> src/stream/src/executor/source/source_executor.rs:91:61
    |
91  |               state_table: core.split_state_store.state_store.clone(),
    |                                                               ^^^^^ method cannot be called due to unsatisfied trait bounds
    |
   ::: src/stream/src/common/table/state_table.rs:96:1
    |
96  | / pub struct StateTableInner<
97  | |     S,
98  | |     SD = BasicSerde,
99  | |     const IS_REPLICATED: bool = false,
100 | |     W = DefaultWatermarkBufferStrategy,
101 | |     const USE_WATERMARK_CACHE: bool = false,
102 | | > where
    | | -
    | | |
    | |_method `clone` not found for this struct
    |   doesn't satisfy `_: Clone`
    |
note: trait bound `<S as risingwave_storage::StateStore>::Local: Clone` was not satisfied
   --> src/stream/src/common/table/state_table.rs:95:10
    |
95  | #[derive(Clone)]
    |          ^^^^^ unsatisfied trait bound introduced in this `derive` macro
    = help: items from traits can only be used if the trait is implemented and in scope
    = note: the following trait defines an item `clone`, perhaps you need to implement it:
            candidate #1: `Clone`
```